src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManager.cs

```diff
@@ -0,0 +1,499 @@
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using LibGit2Sharp;
+using Microsoft.DotNet.Darc.Models.VirtualMonoRepo;
+using Microsoft.DotNet.DarcLib.Helpers;
+using Microsoft.Extensions.Logging;
+
+#nullable enable
+namespace Microsoft.DotNet.DarcLib.VirtualMonoRepo;
+
+public class VmrManager : IVmrManager
+{
+    private const string HEAD = "HEAD";
+
+    // Message shown when synchronizing a single commit
+    private const string SingleCommitMessage =
+        """
+        [{name}] Sync {newShaShort}: {commitMessage}
+
+        Original commit: {remote}/commit/{newSha}
+        """;
+
+    // Message shown when synchronizing multiple commits as one
+    private const string SquashCommitMessage =
+        """
+        [{name}] Sync {oldShaShort} → {newShaShort}
+        Diff: {remote}/compare/{oldSha}..{newSha}
+        
+        From: {remote}/commit/{oldSha}
+        To: {remote}/commit/{newSha}
+        
+        Commits:
+        {commitMessage}
+        """;
+
+    private readonly ILogger<VmrManager> _logger;
+    private readonly IProcessManager _processManager;
+    private readonly IRemoteFactory _remoteFactory;
+    private readonly string _vmrPath;
+    private readonly string _reposPath;
+    private readonly string _tagsPath;
+    private readonly string _tmpPath;
+
+    public IReadOnlyCollection<SourceMapping> Mappings { get; }
+
+    public VmrManager(
+        IProcessManager processManager,
+        IRemoteFactory remoteFactory,
+        ILogger<VmrManager> logger,
+        IReadOnlyCollection<SourceMapping> mappings,
+        string vmrPath,
+        string tmpPath)
+    {
+        _logger = logger;
+        _processManager = processManager;
+        _remoteFactory = remoteFactory;
+        _tmpPath = tmpPath;
+        _vmrPath = vmrPath;
+        _reposPath = Path.Join(vmrPath, "src");
+        _tagsPath = Path.Join(_reposPath, ".tags");
+
+        Mappings = mappings;
+    }
+
+    public async Task InitializeVmr(
+        SourceMapping mapping,
+        string? targetRevision,
+        bool ignoreWorkingTree,
+        CancellationToken cancellationToken)
+    {
+        if (File.Exists(Path.Combine(_reposPath, $".{mapping.Name}")))
+        {
+            throw new EmptySyncException($"Repository {mapping.Name} already exists");
+        }
+
+        _logger.LogInformation("Initializing {name}", mapping.Name);
+
+        string clonePath = await PrepareClone(mapping);
+        string patchPath = GetPatchFilePath(mapping);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        using var clone = new Repository(clonePath);
+        var commit = GetCommit(clone, targetRevision is null || targetRevision == HEAD ? null : targetRevision);
+
+        await CreatePatch(mapping, clonePath, Constants.EmptyGitObject, commit.Id.Sha, patchPath);
+        cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Should we also pass this token to other async methods? `CreatePatch` has two awaits, so it's possible that the token will expire during the first one. I'm not sure what are the best practices with cancellation tokens, need to do more research

Reply:
We intentionally make it so that some calls finish and we can do a graceful termination. E.g. when we call git via `ProcessManager`, we don't want to cancel that (for instance if we kill `git commit` halfway through, it will leave the repository in a weird state, for instance it will leave the `.lock` file behind and further git commands will fail).
We want the graceful termination for:
- Local runs because you wouldn't be able to kill the run exactly in between git operations as those are mostly what is happening.
- When AzDO build times out, we still might want to push what was comitted and continue the sync next time.

Reply:
Also, the async part itself is not the important when dealing with cancellation tokens (since we also `await` all of this serially)